turbopack/crates/turbo-tasks/src/parallel.rs

```diff
@@ -0,0 +1,535 @@
+//! Parallel for each resp. map running in the current tokio thread pool maintaining turbo tasks and
+//! tracing context.
+//!
+//! This avoid the problem of sleeping threads with mimalloc when using rayon in combination with
+//! tokio. It also avoid having multiple thread pools.
+
+use std::{
+    mem::{ManuallyDrop, transmute},
+    panic,
+    sync::{Arc, LazyLock},
+    thread::available_parallelism,
+};
+
+use tokio::{
+    runtime::Handle,
+    task::{JoinHandle, block_in_place},
+};
+use tracing::{Instrument, Span};
+
+use crate::{
+    TurboTasksApi,
+    manager::{try_turbo_tasks, turbo_tasks_try_scope},
+};
+
+/// Calculates a good chunk size for parallel processing based on the number of available threads.
+/// This is used to ensure that the workload is evenly distributed across the threads.
+fn good_chunk_size(len: usize) -> usize {
+    static GOOD_CHUNK_COUNT: LazyLock<usize> =
+        LazyLock::new(|| available_parallelism().map_or(16, |c| c.get() * 4));
+    let min_chunk_count = *GOOD_CHUNK_COUNT;
+    len.div_ceil(min_chunk_count)
+}
+
+/// Context to allow spawning a task with a limited lifetime.
+///
+/// ## Safety
+///
+/// This context must not be dropped before all tasks spawned with it have been awaited.
+struct ProcessInParallelContext<'l, R: Send + 'l> {
+    results: Box<[Option<R>]>,
```

Comment:
I'd rather just make this a `Box<[Mutex<Option<R>>]>` and reduce the amount of `unsafe` code needed (`SendablePtr`). There wouldn't be any lock contention, so I don't think it would matter much for performance.